### PR TITLE
Remove signer hardcoding of SHA-256

### PR DESCRIPTION
**Pull request by malancas**
#### Summary

Closes https://github.com/sigstore/timestamp-authority/issues/304

Currently when signers are created, SHA-256 is always used. A KMS based key may use a different supported algorithm, like SHA-384 or SHA-512, so this hardcoding should be removed.

- Adds a new timestamp-server flag `signer-hash-function` that allows the user to specify a signer hash function. This is used in several different signer schemes
- Adds two new fetch-tsa-certs flags, `intermediate-hash-function` and `leaf-hash-function`, that allow the user to specify the leaf and intermediate key hash functions
- Updates the signer code to handle a specified hash function

### CODE REVIEW

File: pkg/signer/signer.go [hunk 1]
```diff
@@ -36,15 +36,40 @@ const TinkScheme = "tink"
...
+	case "sha384":
+		return crypto.SHA384, elliptic.P384(), nil
+	case "sha512":
+		return crypto.SHA512, elliptic.P521(), nil
+	default:
+		return 0, nil, fmt.Errorf("invalid hash algorithm - must be either sha256, sha384, or sha512")
+	}
+}
+
+func NewCryptoSigner(ctx context.Context, signer, kmsKey, tinkKmsKey, tinkKeysetPath, hcVaultToken, fileSignerPath, fileSignerPasswd, signerHashFunc string) (crypto.Signer, error) {
```
Comment (malancas) [inline pkg/signer/signer.go:52]: I think this function is ready for a refactor but given that this PR already touches over ten files, it can be tackled in a follow up.

File: pkg/signer/file.go [hunk 2]
```diff
@@ -30,26 +30,29 @@ type File struct {
...
 	case *ecdsa.PrivateKey:
-		signer, err := signature.LoadECDSASignerVerifier(pk, crypto.SHA256)
+		signer, err := signature.LoadECDSASignerVerifier(pk, hashFunc)
 		if err != nil {
 			return nil, err
 		}
 		return &File{signer}, nil
 	case ed25519.PrivateKey:
+		if hashFunc != crypto.SHA512 {
+			fmt.Printf("ed25519 only supports SHA512, specified hash func is %s. Using SHA512\n", hashFunc)
```
Comment (malancas) [inline pkg/signer/file.go:54]: Regarding how to handle a hash function that doesn't match the one used by Ed25519, I thought it may make sense to simply inform the user starting the server that the provided algorithm will be disregarded. Thoughts?

File: pkg/signer/tink.go [hunk 3]
```diff
@@ -129,6 +129,9 @@ func KeyHandleToSigner(kh *keyset.Handle) (crypto.Signer, error) {
 		p.PublicKey.X, p.PublicKey.Y = c.ScalarBaseMult(privKey.GetKeyValue())
 		return p, nil
 	case ed25519SignerTypeURL:
+		if hashFunc != crypto.SHA512 {
+			fmt.Printf("Ed25519 only supports SHA512, specified hash func is %s. Using hash function specified by Ed25519\n", hashFunc)
```
Comment (malancas) [inline pkg/signer/tink.go:133]: Regarding how to handle a hash function that doesn't match the one used by Ed25519, I thought it may make sense to simply inform the user that the provided algorithm will be disregarded and move on with setup. Thoughts?

File: pkg/signer/file.go [hunk 4]
```diff
@@ -30,26 +30,29 @@ type File struct {
...
 	case *ecdsa.PrivateKey:
-		signer, err := signature.LoadECDSASignerVerifier(pk, crypto.SHA256)
+		signer, err := signature.LoadECDSASignerVerifier(pk, hashFunc)
 		if err != nil {
 			return nil, err
 		}
 		return &File{signer}, nil
 	case ed25519.PrivateKey:
+		if hashFunc != crypto.SHA512 {
+			fmt.Printf("Ed25519 only supports SHA512, specified hash func is %s. Using hash function specified by Ed25519\n", hashFunc)
```
Comment (haydentherapper) [inline pkg/signer/file.go:54]: Should we err out instead of printing?

File: pkg/signer/memory.go [hunk 5]
```diff
@@ -30,12 +30,25 @@ import (
 	tsx509 "github.com/sigstore/timestamp-authority/pkg/x509"
 )
 
+func getCurveFromSigner(signer crypto.Signer) (elliptic.Curve, error) {
```
Comment (haydentherapper) [inline pkg/signer/memory.go:33]: memory signer hardcodes the signer's curve - https://github.com/sigstore/timestamp-authority/blob/403a0a1fc1944a77ff4c4c3fa528f3d90c518d35/pkg/signer/signer.go#L42 - so this isn't needed

I think this is reasonable, as it's only meant for testing.

File: pkg/signer/signer.go [hunk 6]
```diff
@@ -36,15 +36,40 @@ const TinkScheme = "tink"
...
 	switch signer {
 	case MemoryScheme:
-		sv, _, err := signature.NewECDSASignerVerifier(elliptic.P256(), rand.Reader, crypto.SHA256)
+		hashFunc, curve, err := getHashFuncEllipticCurve(signerHashFunc)
```
Comment (haydentherapper) [inline pkg/signer/signer.go:55]: I'd vote to keep this hardcoded, as it's only meant for testing and should need no additional configuration.

File: pkg/signer/tink.go (hunk 3 above)
Comment (haydentherapper) [reply pkg/signer/tink.go:133]: Given this runs only on setup, I'd prefer we err out rather than allow for misconfiguration.

File: pkg/signer/tink.go [hunk 7]
```diff
@@ -51,7 +51,7 @@ var (
 )
 
 // NewTinkSigner creates a signer by decrypting a local Tink keyset with a remote KMS encryption key
-func NewTinkSigner(ctx context.Context, tinkKeysetPath string, primaryKey tink.AEAD) (crypto.Signer, error) {
+func NewTinkSigner(ctx context.Context, tinkKeysetPath string, primaryKey tink.AEAD, hashFunc crypto.Hash) (crypto.Signer, error) {
```
Comment (haydentherapper) [inline pkg/signer/tink.go:54]: Is hash func needed here? This only builds a signer

**Review by haydentherapper — CHANGES_REQUESTED** [review]
I think we actually need to fix this in the timestamp library, not here. sigstore/sigstore exposes a `SignMessage` interface that takes the hash passed into it, but that's only used by sigstore services. The `crypto.Signer` interface provides `Sign` which takes in a hash function in `crypto.SignerOpts`, but that's only to check that the digest provided matches the hash function expectations (like digest length). `Sign` is what's called in the [pkcs7 library](https://github.com/digitorus/pkcs7/blob/master/sign.go#L373-L404)

The hash function is passed in the timestamp library here - https://github.com/digitorus/timestamp/blob/d542479a242518f315934e8d79c8e077ddbe4990/timestamp.go#L599, and you can see it's hardcoded. I don't think any changes are needed in the pkcs7 library, as it supports other hash algorithms already - https://github.com/digitorus/pkcs7/blob/master/sign.go#L150-L156

so tl;dr - https://github.com/digitorus/timestamp/blob/d542479a242518f315934e8d79c8e077ddbe4990/timestamp.go#L599 needs to be configurable

Leaving up the comments I've typed up already, but they're moot given the above.

**Comment by malancas** [comment]
Closing as the hardcoding bug should be resolved by https://github.com/sigstore/timestamp-authority/pull/452
